fuzzypp/src/cliopts.cpp

```diff
@@ -0,0 +1,112 @@
+#include <iostream>
+
+#include "cliopts.hpp"
+
+namespace fuzzypp::cliopts {
+    const std::optional<CliOptions> 
+    CliOptions::parse_command_line(int argc, char* argv[]) {
+        try {
+            cxxopts::Options options { argv[0], " - options" };
+            options
+                .positional_help("[optional args]")
+                .show_positional_help();
+            
+            options
+                .allow_unrecognised_options()
+                .add_options()
+                ("f,file", "A set of input source files (.c/.cpp).", cxxopts::value<std::vector<std::string>>(), "FILE")
+                ("o,output", "A directory the output should be written to", cxxopts::value<std::string>(), "FILE")
+                ("I", "A set of paths to include on the header search path", cxxopts::value<std::vector<std::string>>(), "DIR")
+                ("include", "A set of files to include on the header search path", cxxopts::value<std::vector<std::string>>(), "FILE")
+                ("D,define", "A set of defined values", cxxopts::value<std::vector<std::string>>(), "NAME[=VALUE]")
+                ("U,undefine", "A set of undefined values", cxxopts::value<std::vector<std::string>>(), "NAME")
+                ("help", "Print help.");
+
+            auto result = options.parse(argc, argv);
+
+            if (result.count("help")) {
+                std::cout << options.help() << std::endl;
+            }
+
+            auto files = extract_vector(result, "file");
+            auto i_paths = extract_vector(result, "I");
+            auto i_files = extract_vector(result, "include");
+            auto defs = extract_vector(result, "define");
+            auto undefs = extract_vector(result, "undefine");
+            auto output = result.count("output") ?
+                result["output"].as<std::string>() : "";
+
+            return std::optional<CliOptions> {
+                CliOptions {
+                    files,
+                    i_files,
+                    i_paths,
+                    defs,
+                    undefs,
+                    output
+                }
+            };
+        } catch (const cxxopts::OptionException& e) {
+            std::cerr << "Error occured whilst parsing command line arguments: " << e.what() << std::endl;
+            return std::optional<CliOptions> {};
+        }
+    }
+
+    inline const std::vector<std::string> 
```

Comment:
Are you sure that's correct? I can't find anything stating this.

https://en.cppreference.com/w/cpp/language/inline
https://isocpp.org/wiki/faq/inline-functions#where-to-put-inline-keyword

Reply:
Yes, how would the compiler _inline_ the function if it can't see it? The compiler can't look across different .cpp, only in .h.  Note that putting an inline function **outside the class body** doesn't mean that it has to be in the .cpp, you can (and should) have those definitions in the .h. Check that same page you linked (see screenshot):

<img width="1077" alt="Screen Shot 2019-10-11 at 11 09 45" src="https://user-images.githubusercontent.com/2905588/66658621-39868c00-ec18-11e9-98e5-2d0f96f88820.png">